src/v/resource_mgmt/memory_sampling.cc

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "resource_mgmt/memory_sampling.h"
+
+#include "resource_mgmt/available_memory.h"
+#include "ssx/future-util.h"
+#include "ssx/sformat.h"
+#include "vlog.h"
+
+#include <seastar/core/future.hh>
+#include <seastar/core/memory.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/core/sstring.hh>
+#include <seastar/util/memory_diagnostics.hh>
+
+#include <fmt/core.h>
+#include <fmt/format.h>
+
+#include <limits>
+#include <vector>
+
+template<>
+struct fmt::formatter<seastar::memory::allocation_site>
+  : fmt::formatter<std::string_view> {
+    template<typename FormatContext>
+    auto
+    format(const seastar::memory::allocation_site& site, FormatContext& ctx) {
+        return fmt::format_to(
+          ctx.out(), "{} {} {}", site.size, site.count, site.backtrace);
+    }
+};
+
+constexpr std::string_view diagnostics_header() {
+    return "Top-N alloc sites - size count stack:";
+}
+
+/// Put `top_n` allocation sites into the front of `allocation_sites`
+static void top_n_allocation_sites(
+  std::vector<ss::memory::allocation_site>& allocation_sites, size_t top_n) {
+    std::partial_sort(
+      allocation_sites.begin(),
+      allocation_sites.begin() + top_n,
+      allocation_sites.end(),
+      [](const auto& lhs, const auto& rhs) { return lhs.size > rhs.size; });
+}
+
+ss::noncopyable_function<void(ss::memory::memory_diagnostics_writer)>
+memory_sampling::get_oom_diagnostics_callback() {
+    // preallocate those so that we don't allocate on OOM
+    std::vector<seastar::memory::allocation_site> allocation_sites(1000);
+    std::vector<char> format_buf(1000);
+
+    return [allocation_sites = std::move(allocation_sites),
+            format_buf = std::move(format_buf)](
+             seastar::memory::memory_diagnostics_writer writer) mutable {
+        auto num_sites = ss::memory::sampled_memory_profile(allocation_sites);
+
+        const size_t top_n = std::min(size_t(10), num_sites);
+        top_n_allocation_sites(allocation_sites, top_n);
+
+        writer(diagnostics_header());
+        writer("\n");
+
+        for (size_t i = 0; i < top_n; ++i) {
+            auto bytes_written = fmt::format_to_n(
```

Comment:
So this will just truncate is 1000 is not enough for 1 site, right?

Reply:
Correct yes

Reply:
Seems like something a unit test might be useful to check since this is type of restriction which tends to get broken by a change elsewhere (e.g., change to output format, including more info, etc), and not detected until you suddenly need it.